Library/OcMainLib/OpenCoreUefiUnloadDrivers.c

```diff
@@ -0,0 +1,352 @@
+                  EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                  );
+  if (!EFI_ERROR (Status)) {
+    BestLang = GetBestLanguageForDriver (CompNameStruct->SupportedLanguages, Language, FALSE);
+    Status   = CompNameStruct->GetDriverName (CompNameStruct, BestLang, &RetVal);
+    if (BestLang != NULL) {
+      FreePool (BestLang);
+    }
+
+    if (!EFI_ERROR (Status)) {
+      return (RetVal);
+    }
+  }
+
+  return (NULL);
+}
+
+VOID
+ReportImageName (
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  )
+{
+  DRIVER_REPORT_CONTEXT  *ReportContext;
+
+  ReportContext = Context;
+
+  OcAsciiPrintBuffer (
+    &ReportContext->FileBuffer,
+    &ReportContext->FileBufferSize,
+    "%s\n",
+    Name
+    );
+}
+
+VOID
+UnloadImageByName (
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  )
+{
+  EFI_STATUS            Status;
+  UINTN                 Index;
+  UNLOAD_IMAGE_CONTEXT  *UnloadContext;
+
+  UnloadContext = Context;
+
+  for (Index = 0; Index < UnloadContext->UnloadNameCount; Index++) {
+    if (StrCmp (UnloadContext->UnloadInfo[Index].UnloadName, Name) == 0) {
+      Status = gBS->UnloadImage (Handle);
+      DEBUG ((
+        EFI_ERROR (Status) ? DEBUG_WARN : DEBUG_INFO,
+        "OC: UnloadImage %s - %r\n",
+        UnloadContext->UnloadInfo[Index].UnloadName,
+        Status
+        ));
+      UnloadContext->UnloadInfo[Index].Unloaded = TRUE; ///< Do not report as 'not found' even if we failed to unload it.
+      break;
+    }
+  }
+}
+
+//
+// Perform action for all driver binding protocols which are
+// present on the same handle as loaded image protocol and
+// where the handle has a usable name.
+//
+// Note: In terms of choosing a name to identify the driver, while the loaded image
+// section name is shorter and corresponds better to the driver's file name, and the
+// firmware volume GUID from the loaded image path identifies the driver more uniquely,
+// these are both not always available for a loaded image (across various firmware),
+// whereas the driver component name normally is.
+//
+VOID
+ProcessAllDrivers (
+  IN VOID        *Context,
+  PROCESS_IMAGE  ProcessImage
+  )
+{
+  EFI_STATUS                 Status;
+  UINTN                      HandleCount;
+  EFI_HANDLE                 *HandleBuffer;
+  EFI_LOADED_IMAGE_PROTOCOL  *LoadedImage;
+  UINTN                      Index;
+  CONST CHAR16               *Name;
+
+  Status = gBS->LocateHandleBuffer (
+                  ByProtocol,
+                  &gEfiDriverBindingProtocolGuid,
+                  NULL,
+                  &HandleCount,
+                  &HandleBuffer
+                  );
+  if (EFI_ERROR (Status)) {
+    return;
+  }
+
+  for (Index = 0; Index < HandleCount; Index++) {
+    Status = gBS->HandleProtocol (
+                    HandleBuffer[Index],
+                    &gEfiLoadedImageProtocolGuid,
+                    (VOID **)&LoadedImage
+                    );
```

Comment:
The other issue is which drivers we try to list (in SysReport) and allow unloading.

We currently only support drivers which have loaded image and driver binding protocol, and a component name. This is only a small subset of the drivers in any given firmware - it does capture the network stack, which is what https://github.com/acidanthera/OpenCorePkg/pull/549 is concerned with, and which seem to be cleanly designed to allow correct unloaded. I believe many/most of the other drivers which this system won't currently list or unload don't support unloading anyway.

Reply:
Correct - DriverBinding drivers are UEFI drivers and must be able to unload. The rest are DXE drivers and may or may not be unloadable, you cannot really know from the outside. Further, as per your other comment, we cannot really identify them very well anyway.